src/Antiforgery/src/Internal/DefaultAntiforgery.cs

```diff
@@ -376,39 +376,40 @@ private IAntiforgeryFeature GetTokensInternal(HttpContext httpContext)
         /// </summary>
         /// <param name="httpContext">The <see cref="HttpContext"/>.</param>
         protected virtual void SetDoNotCacheHeaders(HttpContext httpContext)
-        {
-            // Since antiforgery token generation is not very obvious to the end users (ex: MVC's form tag generates them
-            // by default), log a warning to let users know of the change in behavior to any cache headers they might
-            // have set explicitly.
-            LogCacheHeaderOverrideWarning(httpContext.Response);
-
-            httpContext.Response.Headers[HeaderNames.CacheControl] = "no-cache, no-store";
-            httpContext.Response.Headers[HeaderNames.Pragma] = "no-cache";
-        }
-
-        private void LogCacheHeaderOverrideWarning(HttpResponse response)
         {
             var logWarning = false;
-            if (CacheControlHeaderValue.TryParse(response.Headers[HeaderNames.CacheControl].ToString(), out var cacheControlHeaderValue))
+            
+            if (CacheControlHeaderValue.TryParse(httpContext.Response.Headers[HeaderNames.CacheControl].ToString(), out var cacheControlHeaderValue))
             {
-                if (!cacheControlHeaderValue.NoCache)
+                // If the Cache-Control is already set, override it only if required
+                if (!cacheControlHeaderValue.NoCache || !cacheControlHeaderValue.NoStore)
                 {
                     logWarning = true;
+                    httpContext.Response.Headers[HeaderNames.CacheControl] = "no-cache, no-store";
                 }
             }
+            else
+            {
+                httpContext.Response.Headers[HeaderNames.CacheControl] = "no-cache, no-store";
+            }
 
-            var pragmaHeader = response.Headers[HeaderNames.Pragma];
-            if (!logWarning
-                && !string.IsNullOrEmpty(pragmaHeader)
-                && !string.Equals(pragmaHeader, "no-cache", StringComparison.OrdinalIgnoreCase))
+            var pragmaHeader = httpContext.Response.Headers[HeaderNames.Pragma];
+            if (!logWarning &&
+                !string.IsNullOrEmpty(pragmaHeader) &&
+                !string.Equals(pragmaHeader, "no-cache", StringComparison.OrdinalIgnoreCase))
```

Comment:
You could avoid an extra lookup if we know we're already logging a warning:

```suggestion
            if (!logWarning &&
                response.TryGetValue(HeaderNames.Pragma, out var pragma) &&
                pragma.Count > 0 &&
                !string.Equals(pragma[0], "no-cache", StringComparison.OrdinalIgnoreCase))
```